python/AMI_NOT_PUBLIC_CHECK/AMI_NOT_PUBLIC_CHECK.py

```diff
@@ -0,0 +1,406 @@
+
+Description:
+  Check whether the Amazon Machine Images are not publicly accessible. The rule is NON_COMPLIANT if one or more Amazon Machine Images are publicly accessible.
+
+Trigger:
+  Periodic
+
+Reports on:
+  AWS::::Account
+Rule Parameters:
+  None
+
+Scenarios:
+  Scenario: 1
+    Given: No AMIs with "is-public" parameter set to True
+     Then: Return NOT_APPLICABLE
+  Scenario: 2
+    Given: One or more AMIs with is-public parameter set to True
+     Then: Return NON_COMPLIANT with Annotation containing AMIIDs
+'''
+
+import json
+import datetime
+import botocore
+import boto3
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::::Account'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+#############
+# Main Code #
+#############
+
+# Generates non-compliant response
+def generate_evaluation_list(images, event):
+    evaluations = []
+    # Looping through all available images
+    for image in images:
+        evaluation = build_evaluation(
+            image['ImageId'],
+            "NON_COMPLIANT",
+            event,
+            resource_type=DEFAULT_RESOURCE_TYPE,
+            annotation="Amazon Machine Image Id: {} is public".format(image['ImageId'])
+        )
+        evaluations.append(evaluation)
+    return evaluations
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    """Form the evaluation(s) to be return to Config Rules
+    Return either:
+    None -- when no result needs to be displayed
+    a string -- either COMPLIANT, NON_COMPLIANT or NOT_APPLICABLE
+    a dictionary -- the evaluation dictionary, usually built by build_evaluation_from_config_item()
+    a list of dictionary -- a list of evaluation dictionary , usually built by build_evaluation()
+    Keyword arguments:
+    event -- the event variable given in the lambda handler
+    configuration_item -- the configurationItem dictionary in the invokingEvent
+    valid_rule_parameters -- the output of the evaluate_parameters() representing validated parameters of the Config Rule
+    Advanced Notes:
+    1 -- if a resource is deleted and generate a configuration change with ResourceDeleted status, the Boilerplate code will put a NOT_APPLICABLE on this resource automatically.
+    2 -- if a None or a list of dictionary is returned, the old evaluation(s) which are not returned in the new evaluation list are returned as NOT_APPLICABLE by the Boilerplate code
+    3 -- if None or an empty string, list or dict is returned, the Boilerplate code will put a "shadow" evaluation to feedback that the evaluation took place properly
+    """
+
+    ###############################
+    # Add your custom logic here. #
+    ###############################
+
+    ec2_client = get_client('ec2', event)
+    public_ami_result = ec2_client.describe_images(
```

Comment:
describe_images can paginate. Need to retrieve everything.

Reply:
I do not see pagination in the boto3 API: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_images

DescribeImages API -> https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeImages.html

Could you please guide me to the correct link so that I can incorporate that in the code?

Reply:
Ok, then we will do without :)